src/hotspot/share/gc/parallel/parallelScavengeHeap.cpp

```diff
@@ -549,8 +549,26 @@ void ParallelScavengeHeap::collect(GCCause::Cause cause) {
     return;
   }
 
-  VM_ParallelGCSystemGC op(gc_count, full_gc_count, cause);
-  VMThread::execute(&op);
+  while (true) {
+    VM_ParallelGCSystemGC op(gc_count, full_gc_count, cause);
+    VMThread::execute(&op);
+
+    if (!VM_ParallelGCSystemGC::is_cause_full(cause) || op.full_gc_succeeded()) {
```

Comment:
Why `full_gc_succeeded` here? Doesn't the following `full_gc_count != total_full_collections` achieve the same purpose?

Reply:
True, but this is not a perf-critical method, be consistent (with other gcs) is more important, IMO.

Reply:
Only Serial doesn't have this early return, and only reason I didn't add it is because PR was getting too big and serial required a lot more changes to have this implemented. So i think It will added later for Serial GC.

Reply:
Please file an issue for this.